content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md

```diff
@@ -0,0 +1,198 @@
+buy a micro-USB to USB-A adapter to hook up most other devices.
+
+You do still need a micro SD card. While there are versions of the Rock with 
+built-in flash, it's small and not meant for booting an OS. Note that the size of the built-in flash
+is listed in giga*bits*. The 4Gb version is actually 512 mega*bytes*. So get an 
+SD card, preferably one with an A1 rating, which specifies that it works well 
+for random access patterns. The traditional SD card ratings (like "class 10" or 
+"UHS-I") specify sequential read and write, which is fine for cameras, but not 
+fine for running an operating system.
+
+The big thing in the Rock's favor is a quad core ARM processor. While it's not 
+going to set any Bitcoin mining records, I've found multicore 
+processors are handy for hardware projects. On a single core processor running 
+Linux, your program can get interrupted by something else at any time. This 
+means that a program running on a mere 16MHz Arduino may run better than a 
+1GHz Raspberry Pi. A quad core processor mitigates this, since secondary 
+programs have 3 other cores to work with.
+
+The Rock is powered over its USB-C port. I found attaching it to a 
+laptop USB port provided enough power to get going. This is also handy because 
+you can run `adb shell` to connect to the Rock's OS. "ADB" is the 
+Android Debug Bridge, and while the Rock does support running Android, I used 
+a Ubuntu Bionic image. Yes, `adb` did still work with Ubuntu installed. This is great for 
+headless installs, since you won't have networking available at first boot.
+
+Since it's running Ubuntu, we have `perl` from the start:
+
+```
+$ perl -v:
+
+This is perl 5, version 26, subversion 1 (v5.26.1) built for aarch64-linux-gnu-thread-multi
+(with 67 registered patches, see perl -V for more detail)
+...
+```
+
+What you won't have is an equivalent to [RPi::WiringPi]({{< mcpan "RPi::WiringPi">}})
+to control GPIO ("General Purpose Input/Output", things like blinking an LED or 
+taking input from a switch). That's OK, because it's Linux, and Linux provides a 
+simple sysfs interface to access GPIO pins. As long as there's a GPIO driver 
+for the board, you can control the pins.
+
+The Rock has two headers, one of which is color coded and has the main pins you want to work with in most projects. Note 
+that there are a few variations of the pinout depending on the version of 
+the board you get. The version should be printed on the board (I received a 
+V12 board), so make sure you're looking at the right pinout for your board. 
+See the [GPIO header documentation](https://wiki.radxa.com/RockpiS/hardware/rockpiS#gpio).
+
+One thing to keep in mind is that the pins are specified as 3.3V input. That means you don't want to put 5V 
+into them. Feeding it 5V might cause glitches, or could break the board 
+completely.
+
+The sysfs interface from Linux is documented at:
+
+https://www.kernel.org/doc/html/latest/admin-guide/gpio/sysfs.html
```

Comment:
Make the a link, with a sentence like "Check out the [sysfs interface](...) ..."